patches/sitecustomize.py

```diff
@@ -1,23 +1,29 @@
 import os
-
+from kaggle_gcp import get_integrations
 kaggle_proxy_token = os.getenv("KAGGLE_DATA_PROXY_TOKEN")
+kernel_integrations_var = os.getenv("KAGGLE_KERNEL_INTEGRATIONS")
+
 bq_user_jwt = os.getenv("KAGGLE_BQ_USER_JWT")
 if kaggle_proxy_token or bq_user_jwt:
     from google.auth import credentials
     from google.cloud import bigquery
     from google.cloud.bigquery._http import Connection
     # TODO: Update this to the correct kaggle.gcp path once we no longer inject modules
     # from the worker.
-    from kaggle_gcp import PublicBigqueryClient
+    from kaggle_gcp import PublicBigqueryClient, KaggleKernelCredentials
 
     def monkeypatch_bq(bq_client, *args, **kwargs):
         data_proxy_project = os.getenv("KAGGLE_DATA_PROXY_PROJECT")
         specified_project = kwargs.get('project')
         specified_credentials = kwargs.get('credentials')
-        if specified_project is None and specified_credentials is None:
+        kernel_integrations = get_integrations()
+        if specified_project is None and specified_credentials is None and not kernel_integrations.has_bigquery():
             print("Using Kaggle's public dataset BigQuery integration.")
             return PublicBigqueryClient(*args, **kwargs)
+
         else:
+            if specified_credentials is None:
+                kwargs['credentials'] = KaggleKernelCredentials()
```

Comment:
I see that the approach we've taken here is to use the monkey patching for default credentials.

Just wondering if you considered adding a check in the _default application credentials_ instead: 
https://github.com/googleapis/google-auth-library-python/blob/master/google/auth/_default.py#L298
In that code, we have checks for GAE, GCE, etc.

The reason I ask: we'll have to use similar credentials for GCS integration, and if default credentials are taken care of generically, we will not need to solve the problem of passing credentials object again. 

For example, what if we use exactly the same mechanism you are adding now - see if BQ integration is enabled - but instead if it is enabled, just don't create KaggleKernelCredentials() here and allow default app credentials logic to run, which in turn may use  KaggleKernelCredentials().

Lmk what you think?

Reply:
I considered doing this, but realized we'd need to monkeypatch that default() function because there isn't yet a way to "register" another default checker (https://github.com/googleapis/google-auth-library-python/blob/master/google/auth/_default.py#L298) and monkeypatching that seemed more intrusive than the current approach. What do you think about leaving this as is currently, and I can file a feature request with google-auth-library-python to allow for a more modular checker list (ie. registering checkers). Or do you think monkeypatching the default() function is fine?
There's also the issue: if we use KaggleKernelCredentials as the default credentials then how will we differentiate between if the library requesting is GCS, BQ, etc? Since we will use different secrets/"target"/accounts/auth tokens for each right?

Reply:
I was actually proposing to add the new checker directly to _defult.py (i.e. to send a PR to google-auth-library-python, or at least to reach out to the maintainers - they are from our organization). 

I had in mind something like the following (similar to `def _get_gae_credentials():`:
```
def _get_kaggle_credentials():
    try:
        from kaggle_gcp import KaggleKernelCredentials
    except ImportError:
        return None, None
    ....
```
(i.e., no monkey-patching). 

Regarding how to differentiate - we'll need to think about it, I don't immediately know how it would work, but it seems doable? I imaging there will be some kernels that would try to access BigQuery **and** GCS in the same kernel. I'm happy to brainstorm over chat tomorrow if you have time?

Reply:
I like the idea of using default application credentials as well. Ilya and I spoke about this offline, and I think we can move on with what we have here, but I've reached out to google client library folk to discuss best way to do this more "natively" (ie. either sending a PR to their code base as Ilya proposed, or some other registering mechanism). Once they get back to me we can look into doing this better.

Reply:
👍 Great! Thx for reaching out to the maintainers of the library.

Reply:
Can we get this approved until then, I don't it's a blocker?

Reply:
Sure, this is not a blocker in my opinion as well. LGTM. 
I assume you wanted Seb and Sam to take a look as well since you added them to the review.

Reply:
Sounds good, thanks! I'll hold off submitting until one of them reviews, but any reason not to approve PR?

Reply:
Resolved: the review button disappeared for me, but after you re-requested it worked.